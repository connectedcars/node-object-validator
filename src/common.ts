import { RequiredFail, ValidationFailure, ValidationsError } from './errors'

// https://stackoverflow.com/questions/51651499/typescript-what-is-a-naked-type-parameter
// https://2ality.com/2019/07/testing-static-types.html
// Wrapping the types in an tuple force a specific type instead of allow any in the union
export type AssertEqual<T, Expected> = [T, Expected] extends [Expected, T] ? true : false

export type ReturnEqual<T, C> = [T, C] extends [C, T] ? C : never

// https://github.com/microsoft/TypeScript/issues/41746
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BaseObject = Record<string, any>

// TODO: Give better name
export function isValidType<T>(value: unknown, errors: ValidationFailure[]): value is T {
  return errors.length === 0
}

export type CodeGenResult = [{ [key: string]: unknown }, string[], string[]]

export type ValidatorBaseOptions = {
  optimize?: boolean
  required?: boolean
  nullable?: boolean
  earlyFail?: boolean
}

/**
 * @typedef ValidatorOptions
 * @property {boolean} [earlyFail=false] Stop validation on first failure (default: false)
 * @property {boolean} [optimize=true] Generate an optimized function for doing the validation (default: true)
 */
export interface ValidatorOptions {
  earlyFail?: boolean
  optimize?: boolean
}

export interface ValidateOptions {
  earlyFail?: boolean
  optimized?: boolean
}

export interface ValidatorExportOptions {
  language?: string
  types?: boolean
}

export function isValidator(value: unknown): value is ValidatorBase {
  if (value instanceof ValidatorBase) {
    return true
  }
  return false
}

export function generateOptionsString(options: ValidatorBaseOptions, defaults: Required<ValidatorBaseOptions>): string {
  const selectedOptions: string[] = []
  if (options.required !== undefined && options.required !== defaults.required) {
    selectedOptions.push(`required: ${options.required}`)
  }
  if (options.nullable !== undefined && options.nullable !== defaults.nullable) {
    selectedOptions.push(`nullable: ${options.nullable}`)
  }
  if (options.earlyFail !== undefined && options.earlyFail !== defaults.earlyFail) {
    selectedOptions.push(`earlyFail: ${options.earlyFail}`)
  }
  if (options.optimize !== undefined && options.optimize !== defaults.optimize) {
    selectedOptions.push(`optimize: ${options.optimize}`)
  }
  return selectedOptions.length > 0 ? `{ ${selectedOptions.join(', ')} }` : ''
}

export abstract class ValidatorBase<T = unknown> {
  public tsType!: T
  public required: boolean
  public nullable: boolean
  public earlyFail: boolean

  protected codeGenId = 1
  protected optimizedValidate: ((value: unknown, context?: string) => ValidationFailure[]) | null
  protected optionsString: string

  public constructor(options?: ValidatorBaseOptions) {
    const defaults = { required: true, nullable: false, earlyFail: false, optimize: true }
    const mergedOptions = { ...defaults, ...options }

    this.optionsString = options ? generateOptionsString(options, defaults) : ''

    this.required = mergedOptions.required
    this.earlyFail = mergedOptions.earlyFail
    this.nullable = mergedOptions.nullable
    this.optimizedValidate = null
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public AssertType<C extends T, X extends AssertEqual<T, C>>(): void {
    // Only used for type assertions
  }

  public isValid<C extends T>(obj: unknown): obj is ReturnEqual<T, C> {
    const errors = this.validate(obj)
    return errors.length === 0
  }

  public isType<C extends T>(obj: unknown, errors: ValidationFailure[]): obj is ReturnEqual<T, C> {
    return errors.length === 0
  }

  public cast<C extends T>(obj: unknown): ReturnEqual<T, C> {
    const errors = this.validate(obj)
    if (this.isType<C>(obj, errors)) {
      return obj
    } else {
      throw new ValidationsError('One of more validations failed', errors)
    }
  }

  public validate(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    if (options?.optimized !== false && this.optimizedValidate !== null) {
      return this.optimizedValidate(value, context)
    }
    if (this.nullable && value === null) {
      return []
    }
    if (value === undefined) {
      return this.required ? [new RequiredFail(`Is required`, value, context)] : []
    }
    return this.validateValue(value, context, { earlyFail: this.earlyFail, optimized: false, ...options })
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, { key: (context ? \`\${context}['${context}']\` : \`${context}\`) }` : ', context'
    const validatorName = `validator${id()}`
    const declarations = [`const ${validatorName} = ${validatorRef}`]
    // prettier-ignore
    const code: string[] = [
      ...this.nullCheckWrap([
        `  errors.push(...${validatorName}.validateValue(${valueRef}${contextStr}))`,
      ],valueRef, contextStr),
      ...(this.earlyFail || earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  protected nullCheckWrap(code: string[], valueRef: string, contextStr: string, skip = false): string[] {
    if (skip) {
      return code
    }

    const result: string[] = []
    if (this.required) {
      if (this.nullable) {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(`  if (${valueRef} !== null) {`)
        result.push(...code)
        result.push(`  }`)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      }
    } else {
      if (this.nullable) {
        result.push(`if (${valueRef} != null) {`)
        result.push(...code)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`}`)
      }
    }
    return result
  }

  protected optimize(schema?: unknown): void {
    const [imports, declarations, code] = this.codeGen(
      'value',
      'schema',
      () => {
        return this.codeGenId++
      },
      undefined,
      this.earlyFail
    )
    const functionBody = [
      ...Object.keys(imports).map(i => `const ${i} = imports['${i}']`),
      ...declarations,
      `return (value, context) => {`,
      `  const generatedFunction = true`,
      `  const errors = []`,
      ...code.map(l => `  ${l}`),
      `  return errors`,
      `}`
    ].join('\n')

    try {
      const functionGenerator = new Function('imports', 'schema', functionBody)
      const validateFunction = functionGenerator(imports, {
        schema,
        validateValue: this.validateValue.bind(this)
      })
      this.optimizedValidate = validateFunction
    } catch (e) {
      throw new Error(`Failed to compile optimized function(${e.message}):\n${functionBody}`)
    }
  }

  public abstract toString(options?: ValidatorExportOptions): string

  protected abstract validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[]
}
