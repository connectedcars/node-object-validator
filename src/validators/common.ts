import { ValidationErrorContext } from '../errors'

export type Schema = {
  [key: string]: Validator
}

export interface Validator {
  schema?: Schema
  validate(value: unknown, context?: ValidationErrorContext): Error | null
}
