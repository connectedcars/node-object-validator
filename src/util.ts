import { SupportedLanguages, ValidatorBase, ValidatorExportOptions } from './common'

export function toSnakeCase(str: string): string {
  // Check if the entire string is capitalized
  if (str === str.toUpperCase()) {
    return str
  }

  // Convert to snake case if the string is not fully capitalized
  return str.replace(/[A-Z]/g, (letter, index) => {
    return index === 0 ? letter.toLowerCase() : '_' + letter.toLowerCase()
  })
}

export function toPascalCase(str: string): string {
  // Check if the entire string is capitalized
  if (str === str.toUpperCase()) {
    return str
  }

  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function addTypeDef(typeName: string, typeValue: string, record: Record<string, string>): void {
  // Skip redefining stuff, so if you have a tagged union data object, it doesn't get redefined later on if it's used (without the parent, the key would get added in again)
  if (record[typeName] === undefined) {
    record[typeName] = typeValue
  }
}

const rustTypeNameRegex = /^(?:[A-Za-z_][A-Za-z0-9_]*)(?:::[A-Za-z_][A-Za-z0-9_]*)*$/
export function isValidRustTypeName(typeName: string): boolean {
  return rustTypeNameRegex.test(typeName)
}

export function validateRustTypeName(typeName: string, context: ValidatorBase): void {
  if (isValidRustTypeName(typeName) === false) {
    throw new Error(`validateRustTypeName(): Invalid Rust Type Name: '${typeName}'. In: ${context.toString()}`)
  }
}

export function decoratorsString(
  validator: ValidatorBase,
  typeLanguage: SupportedLanguages,
  unionKey: string | undefined
): string {
  if (typeLanguage === 'rust') {
    const options: SerdeDecoratorsOptions = {
      copyable: validator.copyable,
      hashable: validator.hashable,
      comparable: validator.comparable,
      partialComparable: validator.partialComparable,
      defaultable: validator.defaultable,
      unionKey: unionKey,
      renameAll: 'camelCase'
    }
    const serdeStr = serdeDecorators(options)
      .map(decorator => decorator + '\n')
      .join('')
    return serdeStr
  } else {
    throw new Error(`decoratorsString() called with unsupported language: ${typeLanguage}`)
  }
}

export interface SerdeDecoratorsOptions {
  comparable: boolean
  partialComparable: boolean
  hashable: boolean
  defaultable: boolean
  copyable: boolean
  unionKey?: string
  renameAll?: string
}
export function serdeDecorators(options: SerdeDecoratorsOptions): string[] {
  const decorators = []

  const deriveMacros = [`Serialize`, `Deserialize`, `Debug`, `Clone`]
  if (options.partialComparable) {
    deriveMacros.push(`PartialEq`)
  }
  if (options.comparable) {
    if (!options.partialComparable) {
      deriveMacros.push(`PartialEq`)
    }
    deriveMacros.push(`Eq`)
  }
  if (options.hashable === true) {
    deriveMacros.push(`Hash`)
  }
  if (options.defaultable === true) {
    deriveMacros.push(`Default`)
  }
  if (options.copyable === true) {
    deriveMacros.push(`Copy`)
  }

  decorators.push(`#[derive(${deriveMacros.join(', ')})]`)

  if (options.renameAll !== undefined) {
    decorators.push(`#[serde(rename_all = "${options.renameAll}")]`)
  }

  if (options.unionKey !== undefined) {
    decorators.push(`#[serde(tag = "${options.unionKey}")]`)
  }

  return decorators
}

export function generateRustTypes(validators: ValidatorBase[], inputOptions?: ValidatorExportOptions): string {
  const typeDefinitions: Record<string, string> = {}

  const options: ValidatorExportOptions = {
    ...inputOptions,
    types: true,
    language: 'rust',
    typeDefinitions
  }

  for (const validator of validators) {
    // The type definitions gets added to the hashmap
    validator.toString(options)
  }

  let importContent = ``
  let shouldImportDateTime = false
  let shouldImportHashMap = false

  // Alphabetical order: chrono, serde, hashmap
  let typeContent = ``
  for (const value of Object.values(typeDefinitions)) {
    if (value.includes('DateTime')) {
      shouldImportDateTime = true
    }
    if (value.includes('HashMap')) {
      shouldImportHashMap = true
    }
    typeContent += value
  }

  // Chrono (time)
  if (shouldImportDateTime) {
    importContent += `use chrono::{DateTime, Utc};\n`
  }
  // Serde
  importContent += `use serde::{Deserialize, Serialize};\n`
  // Hashmap (Record)
  if (shouldImportHashMap) {
    importContent += `use std::collections::HashMap;\n`
  }
  importContent += `\n`

  const content = importContent + typeContent
  return content
}
