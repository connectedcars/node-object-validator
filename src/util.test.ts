import {
  OptionalArray,
  OptionalBoolean,
  OptionalString,
  RequiredBoolean,
  RequiredDate,
  RequiredDateTime,
  RequiredExactString,
  RequiredFloat,
  RequiredInteger,
  RequiredObject,
  RequiredRecord,
  RequiredString,
  RequiredTuple,
  RequiredUnion,
  RequiredUnixDateTime
} from '.'
import { ValidatorBase } from './common'
import { generateRustTypes, toPascalCase, toSnakeCase } from './util'

describe('toSnakeCase', () => {
  it('lower case first', () => {
    const res = toSnakeCase('katteKilling')
    expect(res).toEqual('katte_killing')
  })

  it('upper case first', () => {
    const res = toSnakeCase('KatteKilling')
    expect(res).toEqual('katte_killing')
  })
})

describe('toPascalCase', () => {
  it('lower case first', () => {
    const res = toPascalCase('katteKilling')
    expect(res).toEqual('KatteKilling')
  })

  it('upper case first', () => {
    const res = toPascalCase('KatteKilling')
    expect(res).toEqual('KatteKilling')
  })
})

describe('generateRustTypes', () => {
  it('write to file', () => {
    const tabbyValidator = new RequiredObject({
      type: new RequiredExactString('tabby'),
      weight: new RequiredFloat(),
      age: new RequiredInteger(0)
    })
    const tuxedoValidator = new RequiredObject({
      type: new RequiredExactString('tuxedo'),
      weight: new RequiredFloat(),
      age: new RequiredInteger(0)
    })
    const maineCoonValidator = new RequiredObject({
      type: new RequiredExactString('maineCoon'),
      weight: new RequiredFloat(),
      age: new RequiredInteger(0),
      furVariant: new RequiredString()
    })

    const catValidator = new RequiredUnion([tabbyValidator, tuxedoValidator, maineCoonValidator], { typeName: 'Cat' })

    const externalTupleValidator = new RequiredTuple([new RequiredInteger(0, 255), new RequiredInteger(0, 255)], {
      typeName: 'ExternalTuple'
    })

    const externalInterfaceValidator = new RequiredUnion(
      [
        new RequiredExactString('CAN0'),
        new RequiredExactString('CAN1'),
        new RequiredExactString('CAN2'),
        new RequiredExactString('VCAN0'),
        new RequiredExactString('VCAN1'),
        new RequiredExactString('VCAN2'),
        new RequiredObject({
          PINS: new RequiredTuple([new RequiredInteger(0, 255), new RequiredInteger(0, 255)])
        }),
        new RequiredObject({
          FAKEEXTTUPLE: externalTupleValidator
        }),
        new RequiredObject({
          FAKEVALUE: new RequiredInteger(0, 255)
        })
      ],
      { typeName: 'ExternalInterface' }
    )

    const transportValidator = new RequiredUnion(
      [new RequiredExactString('ISOTP'), new RequiredExactString('TP2'), new RequiredExactString('ISOTPNOPAD')],
      { typeName: 'Transport' }
    )

    const requestDIDValidator = new RequiredObject(
      {
        adaption_setup: new OptionalArray(new RequiredString()),
        adaption_teardown: new OptionalArray(new RequiredString()),
        service_and_did: new RequiredString(),
        interface: externalInterfaceValidator,
        application: new RequiredUnion(
          [new RequiredExactString('OBD'), new RequiredExactString('UDS'), new RequiredExactString('KWP2000')],
          { typeName: 'Application' }
        ),

        transport: transportValidator,
        session_type: new OptionalString(),
        use_functional_addressing: new OptionalBoolean(),
        tx_id: new RequiredString(),
        rx_id: new OptionalString(),
        tx_local_id: new OptionalString(),
        rx_local_id: new OptionalString(),
        raw: new RequiredBoolean()
      },
      { typeName: 'RequestDID' }
    )

    const validators: ValidatorBase[] = [catValidator, requestDIDValidator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  it('datetime used', () => {
    const validator = new RequiredObject(
      {
        unix: new RequiredUnixDateTime(),
        date: new RequiredDate(),
        dateTime: new RequiredDateTime()
      },
      { typeName: 'RustType' }
    )

    const validators: ValidatorBase[] = [validator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  it('hashmap used', () => {
    const validator = new RequiredObject(
      {
        hashBoi: new RequiredRecord(new RequiredBoolean())
      },
      { typeName: 'RustType' }
    )

    const validators: ValidatorBase[] = [validator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  it('datetime & hashmap used', () => {
    const validator = new RequiredObject(
      {
        dateTime: new RequiredDateTime(),
        hashBoi: new RequiredRecord(new RequiredBoolean())
      },
      { typeName: 'RustType' }
    )

    const validators: ValidatorBase[] = [validator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  it('optional', () => {
    const validator = new RequiredObject(
      {
        valueA: new OptionalBoolean()
      },
      { typeName: 'RustType' }
    )

    const validators: ValidatorBase[] = [validator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  it('overwrite derive macro', () => {
    const validator = new RequiredObject(
      {
        valueA: new RequiredBoolean()
      },
      { typeName: 'RustType', deriveMacro: ['Serialize', 'Deserialize', 'Clone'] }
    )

    const validators: ValidatorBase[] = [validator]
    const types = generateRustTypes(validators)
    expect(types).toMatchSnapshot()
  })

  // #[serde(rename_all = "camelCase")]
  // #[serde(tag = "type")]
  // enum MyEnum {
  // Gives us:
  // Error("cannot serialize tagged newtype variant MyEnum::VariantA containing a boolean", line: 0, column: 0)'
  // VariantA(bool),
  // Which we can't represent, so this catches a similar case
  it('Err, Tagged union, invalid element', () => {
    const validator = new RequiredUnion(
      [
        new RequiredObject({ type: new RequiredExactString('ja'), bingo: new RequiredInteger() }),
        new RequiredBoolean()
      ],
      {
        typeName: 'RustType'
      }
    )

    const validators: ValidatorBase[] = [validator]

    expect(() => {
      generateRustTypes(validators)
    }).toThrow(`Members of the Union(non tagged) are not an ExactString`)
  })
})
