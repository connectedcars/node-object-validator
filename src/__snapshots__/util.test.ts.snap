// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRustTypes datetime & hashmap used 1`] = `
"use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RustType {
    pub date_time: DateTime<Utc>,
    pub hash_boi: HashMap<String, bool>,
}

"
`;

exports[`generateRustTypes datetime used 1`] = `
"use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RustType {
    #[serde(with = \\"chrono::serde::ts_seconds\\")]
    pub unix: DateTime<Utc>,
    pub date: DateTime<Utc>,
    pub date_time: DateTime<Utc>,
}

"
`;

exports[`generateRustTypes hashmap used 1`] = `
"use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RustType {
    pub hash_boi: HashMap<String, bool>,
}

"
`;

exports[`generateRustTypes optional 1`] = `
"use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RustType {
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub value_a: Option<bool>,
}

"
`;

exports[`generateRustTypes overwrite derive macro 1`] = `
"use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RustType {
    pub value_a: bool,
}

"
`;

exports[`generateRustTypes write to file 1`] = `
"use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct TabbyData {
    pub weight: f64,
    pub age: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct TuxedoData {
    pub weight: f64,
    pub age: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct MaineCoonData {
    pub weight: f64,
    pub age: u64,
    pub fur_variant: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
#[serde(tag = \\"type\\")]
pub enum Cat {
    Tabby(TabbyData),
    Tuxedo(TuxedoData),
    MaineCoon(MaineCoonData),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct ExternalTuple(u8, u8);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub enum ExternalInterface {
    CAN0,
    CAN1,
    CAN2,
    VCAN0,
    VCAN1,
    VCAN2,
    PINS(u8, u8),
    FAKEEXTTUPLE(ExternalTuple),
    FAKEVALUE(u8),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub enum Application {
    OBD,
    UDS,
    KWP2000,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub enum Transport {
    ISOTP,
    TP2,
    ISOTPNOPAD,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = \\"camelCase\\")]
pub struct RequestDID {
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub adaption_setup: Vec<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub adaption_teardown: Vec<String>,
    pub service_and_did: String,
    pub interface: ExternalInterface,
    pub application: Application,
    pub transport: Transport,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub session_type: Option<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub use_functional_addressing: Option<bool>,
    pub tx_id: String,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub rx_id: Option<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub tx_local_id: Option<String>,
    #[serde(skip_serializing_if = \\"Option::is_none\\")]
    pub rx_local_id: Option<String>,
    pub raw: bool,
}

"
`;
